File: Design-Patterns-Using-CSharp/Template/TemplatePatternOrder.cs
using System;

abstract class OrderPackingTemplate
{
    // The template method
    public void PackProduct()
    {
        GetProduct();
        AddProductToBox();
        Delivery();
    }

    // Primitive operation: GetProduct
    protected void GetProduct()
    {
        Console.WriteLine("Get the product from the shelf.");
    }

    // Primitive operation: AddProductToBox
    protected void AddProductToBox()
    {
        Console.WriteLine("Put the product inside the box.");
    }

    // Primitive operation: Delivery
    protected abstract void Delivery();
}

class OnlineOrderPacking : OrderPackingTemplate
{
    protected override void Delivery()
    {
        Console.WriteLine("Add delivery address slip and ship.");
    }
}

class StoreOrderPacking : OrderPackingTemplate
{
    protected override void Delivery()
    {
        Console.WriteLine("Add thanks message to the box and deliver to the customer.");
    }
}

// Client code
class TemplatePatternOrder
{
    public static void Main(string[] args)
    {
        OnlineOrderPacking onlineOrder = new OnlineOrderPacking();
        onlineOrder.PackProduct();

        Console.WriteLine();

        StoreOrderPacking storeOrder = new StoreOrderPacking();
        storeOrder.PackProduct();
    }
}

File: Design-Patterns-Using-Java/Template/TemplatePatternOrder.java
import java.io.*;

abstract class OrderPackingTemplate {
    
    final void packProduct() {
        getProduct();
        addProductToBox();
        delivery();
    }

    void getProduct() {
        System.out.println("Get the product from the shelf.");
    }

    void addProductToBox() {
        System.out.println("Put the product inside the box.");
    }

    abstract void delivery();
}

class OnlineOrderPacking extends OrderPackingTemplate {
    @Override
    void delivery() {
        System.out.println("Add delivery address slip and ship.");
    }
}

class StoreOrderPacking extends OrderPackingTemplate {
    @Override
    void delivery() {
        System.out.println("Add thanks message to the box and deliver to the customer.");
    }
}

public class TemplatePatternOrder {
    public static void main(String[] args) {
        OnlineOrderPacking onlineOrder = new OnlineOrderPacking();
        onlineOrder.packProduct();

        System.out.println();

        StoreOrderPacking storeOrder = new StoreOrderPacking();
        storeOrder.packProduct();
    }
}

/*
Get the product from the shelf.
Put the product inside the box.
Add delivery address slip and ship.

Get the product from the shelf.
Put the product inside the box.
Add thanks message to the box and deliver to the customer.
*/